### DISCUSSION TITLE
Testsuite using docker is not super podman friendly

### DISCUSSION THREAD

**Discussion by Romain-Geissler-1A** [libssh2/libssh2#1066] [Ideas]
**Describe the bug**
The current testsuite requires docker, however some people might decide to use podman instead. Also, so logic to retrieve the IP:port of the containerized ssh server is error prone.

**To Reproduce**
These are not strict reproducer commands, but basically I am in such a situation:
 - I have a machine on which both podman and docker are installed. Docker is installed the "classical way", so with a root daemon, not the rootless mode at all. My machine is very firewalled, only very few ports can be reacheable using the machine IP. Podman is also installed the "classical" way, and I am running a non root user, so containers started with podman are rootless.
 - I try to run the build (and later the tests) of libssh2 inside a podman container. The podman container contains a docker client, but not the docker daemon (which resides outside the container).

In such a situation, you can get the test to work, but in many scenarios, it just doesn't work (ie the ip:port to connect to the ssh server is wrong and the connection never happens, making all tests fail).

I tried couple of things:
 - First, don't use docker at all: expose the non-root podman socket inside the container, and fool docker into thinking that the docker daemon it shall use is the podman one (by exporting DOCKER_HOST to the podman socket inside the container). Docker build works, but then the connection doesn't happen. Looking at the code, it seems you detect if you are running inside a container if the cgroup contains the "docker" string," but that's very restrictive, of course podman doesn't do that.
 - I tried exposing the root docker socket inside the container, but then I never really was able to find the right unix user id/group to use inside the image so I would have anything else than "permission denied". Most likely there is a way, but I didn't find it. Anyway, since the test is run from podman, I am pretty sure it doesn't work either (see next points).
 - I tried not using the unix docker socket, but instead the tcp docker socket (as in my case, my server has one, unauthenticated, so absolutely insecure...). Docker build works, but the test don't (no connection happens with the ssh server).
 - Finally I made it work with the above scenario, but using `--network=host` when running the first podman container.

**Expected behavior**

As a developer myself, and knowing a bit how podman/docker works, and how it's possible to do podman inside podman, podman inside docker, docker side by side docker, podman side by side podman, ... and all the possible variants, I think the way you retrieve the ip port is error prone and too restrictive.

For me, you should first allow users to override the "docker" client. Many different implementation of docker exists, trying to mimic the CLI interface. Podman is one of them, but not the only one. There is at least nerdctl, most likely there are others too. Docker is less and less a "standard" as podman rises these days.

Second, about the ip port retrieval, I think you proceed another way. I would rather start a container with openssh server listening on port 22 inside the container, but *WITHOUT* trying to expose this port outside the container (as you really have no idea where it really is running). Then locally, using something like netcat or socat in server mode, bind/listen to port 0 (so a random available ephemeral port is retrieve in the current network namespace, wherever this is), and make netcat/socat exec into a "docker exec container-id netcat" so you have:
 - one netcat in server mode running in the build namespace
 - which spawns and pipes into a netcat running in the test namespace (which is another one)
 - which pipes everything into 127.0.0.1:22 inside the test namespace

this way, you can use ip adress 127.0.0.1 everywhere, you don't need to make any assumption if you are inside or outside a container, using docker or podman, or anything. The only thing you will need to do is "guess" the actual port used when you bound  port 0. It seems socat/netcat doesn't allow this, but it would be easy to write a small C code that acts exactly as netcat/socat does to achieve this and retrieve the actual port behind port 0.

I hope my suggestion makes sense. Generally speaking my experience with containers and CI/CD tells me that port exposition in docker is a nightmare as finding the right ip/port to use in the end to connect to it depends a lot on where the client is relative to the server.

**Comment by dfandrich** [comment]
Would rewriting the container orchestration layer to use [Dagger](https://dagger.io/) solve the portability aspect of this issue? I haven't looked into it in depth, but it sounds like it might provide a cleaner base on which to build the test infrastructure.

**Comment by Romain-Geissler-1A** [comment]
Mmmh I don't know Dagger, but I would skeptical that for the very issue being discussed here it's useful (it can be useful for your CI thought, but that is another matter).

The whole point here is that to test libssh you choose to use an ssh server deployed via docker, and somehow it's being assumed that this docker server has an IP findable and reachable rather easily from the host where libssh is running, and this is what isn't always true (at least in my case). I don't think Dagger won't change much to this problem statement.

If you had the garantee that *BOTH* the libssh environment and the ssh server were run inside containers, assuming they are run inside the same container engine (ie the same podman or the same docker server), then the problem would be trivially fix without using any exposed ports: start one container (for example the ssh server), get its container id (typically with `--cidfile`), then start the second container in the same network namespace (`--network container:$ID_OF_FIRST_CONTAINER`). Yet it's not possible in your case as the libssh is being build on the host (ie *NOT* inside a container) and nothing tells you that what is being built in the host is compatible with what can be run in the docker server (potentially another machine, which another CPU maybe not compatible (even if both have the same arch), and the system libraries of the host will never be the same as the ones you would find in a container).

We faced this use case quite a few times in my company (test binary needs to run in the host (not in a container), while the remote service to be contacted runs in a container. I once opened this request on podman side which has been declined: https://github.com/containers/podman/issues/16801 We use in our company the workaround i have mentionned in this issue. Given that you use ssh specifically, if you are ok to assume that the host has a ssh client (not the libssh one), then instead of using socat you can use the openssh client feature of being able to redirect port over ssh to replace the socat usage of my workaround, and make as is the socket is local to the host (so ip = localhost) while it really is running inside a container.

**Comment by vszakats** [comment]
I was thinking that factoring out the Docker bits from the C session fixture might help with this. Or, creating an alternate fixture in bash with similar functionality. Such script could be much more trivially rewritten or extended for podman or any other containerization (than C). Downside of such script is that it needs an env capable of running it, which brings us to Windows. But maybe this isn't such a big issues nowadays as there are multiple solid ways on Windows to run bash. Pure sshd tests (= `test_ssh2`) required it for quite a while already.
